### ISSUE TITLE
Test failure for entities with service, JPA filtering, required relationship to User and a true DB

### ISSUE THREAD

**Issue by cbornet** [jhipster/generator-jhipster#6600]
##### **Overview of the issue**

The integration tests fail if an entity has service, JPA filtering on, a required relationship to User and use a true DB (MySQL or PosgreSQL in travis)

<details>

```
2017-10-26 19:47:56.795  WARN 8735 --- [           main] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 23505, SQLState: 23505
2017-10-26 19:47:56.796 ERROR 8735 --- [           main] o.h.engine.jdbc.spi.SqlExceptionHelper   : Unique index or primary key violation: "UX_USER_LOGIN_INDEX_6 ON PUBLIC.JHI_USER(LOGIN) VALUES ('johndoe', 2459)"; SQL statement:
insert into jhi_user (created_by, created_date, last_modified_by, last_modified_date, activated, activation_key, email, first_name, image_url, lang_key, last_name, login, password_hash, reset_date, reset_key, id) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) [23505-196]
Tests run: 17, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 2.021 sec <<< FAILURE! - in com.mycompany.myapp.web.rest.TestEntityResourceIntTest
getAllTestEntitiesByUserOneToOneIsEqualToSomething(com.mycompany.myapp.web.rest.TestEntityResourceIntTest)  Time elapsed: 0.114 sec  <<< ERROR!
javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.index.BaseIndex.getDuplicateKeyException(BaseIndex.java:103)
	at org.h2.mvstore.db.MVSecondaryIndex.checkUnique(MVSecondaryIndex.java:231)
	at org.h2.mvstore.db.MVSecondaryIndex.add(MVSecondaryIndex.java:190)
	at org.h2.mvstore.db.MVTable.addRow(MVTable.java:705)
	at org.h2.command.dml.Insert.insertRows(Insert.java:156)
	at org.h2.command.dml.Insert.update(Insert.java:114)
	at org.h2.command.CommandContainer.update(CommandContainer.java:101)
	at org.h2.command.Command.executeUpdate(Command.java:260)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:164)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:150)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:205)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:45)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3003)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3503)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:89)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1435)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1421)
	at sun.reflect.GeneratedMethodAccessor445.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:298)
	at com.sun.proxy.$Proxy173.flush(Unknown Source)
	at com.mycompany.myapp.web.rest.TestEntityResourceIntTest.getAllTestEntitiesByUserOneToOneIsEqualToSomething(TestEntityResourceIntTest.java:281)
```

</details>

See Travis failure [here](https://travis-ci.org/cbornet/generator-jhipster/jobs/293038409#L5975)

Happens since commit https://github.com/jhipster/generator-jhipster/commit/26334ef7add19a17198c5823693016c8d2e633d5

##### **Motivation for or Use Case**

Int tests shouldn't fail OOTB

##### **Reproduce the error**

Change `service` from `no` to `serviceClass` in `TestEntity` and run Travis's `app-mysql`

##### **JHipster Version(s)**

Master and 4.10.1. It's a regression : didn't occur on 4.10.0

##### **JHipster configuration**

```json
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "baseName": "travisPsqlEsNoi18n",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "authenticationType": "session",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": "elasticsearch",
    "useSass": true,
    "buildTool": "maven",
    "enableTranslation": false,
    "enableSocialSignIn": false,
    "rememberMeKey": "f19f0827c622eb9b81f5227a869ccd44932d78eb",
    "testFrameworks": [
      "gatling",
      "protractor"
    ],
    "serverPort": "8080",
    "jhiPrefix": "jhi",
    "travis": true,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "clientPackageManager": "yarn",
    "clientFramework": "angularX"
  }
}
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

[https://github.com/jhipster/generator-jhipster/tree/master/travis/samples/.jhipster/Test*](https://github.com/jhipster/generator-jhipster/tree/master/travis/samples/.jhipster/)

Change `service` from `no` to `serviceClass` in `TestEntity.json`

- [x] Checking this box is mandatory (this is just to show you read everything)

cc @gzsombor

**Comment by cbornet** [comment]
Weird thing is that `TestServiceClass` which also have `serviceClass` and `jpaMetamodelFiltering` doesn't fail so there must be something else...

**Comment by cbornet** [comment]
Might be because `TestEntity` has a required one-to-many relationship with `User`

**Comment by cbornet** [comment]
Confirmed : require relationship to `User` is also part of the issue. If `"relationshipValidateRules": "required"` is removed, the test pass. See https://travis-ci.org/cbornet/generator-jhipster/jobs/293058718

**Comment by gzsombor** [comment]
The new tests are requiring related entities to be created, in this case a new User entity. It's not clear why there are multiple "johndoe" in the db, the changes from the previous tests should be rollbacked...

**Comment by cbornet** [comment]
Indeed ! The problem is that `TestEntity` has multiple relationships to `User` one of which is required. So the required User is created at init and the other ones are created during the filtering tests but using the same login which is set as unique in DB. Hence crash !

There is another filter test failing which is the one with `TestCustomTableName` relationship because  `TestCustomTableName` has a required relationship to `TestEntity` which has a required relationship to `User`. So the `johndoe` user is created twice : first at `TestEntity` test init and then at `TestCustomTableName`+`TestEntity` creation.

What is unclear is why this doesn't fail with H2 ?!

I wonder what happens if A requires B and B requires A. Won't there be a circular creation of the entities in the test init phase ?

**Comment by cbornet** [comment]
I guess a solution for 1/ would be to keep the reference to the user created in init and reuse it in the getAll<Entity>ByUser<Relationship>IsEqualToSomething() methods.

**Comment by cbornet** [comment]
Related to #6521. I guess it will be solved by #6564

**Comment by gzsombor** [comment]
I think, this issue is fixed, so we can close the issue
